pkg/framework/capi_machinesets.go

```diff
@@ -196,3 +200,179 @@ func GetCAPIMachinesFromMachineSet(ctx context.Context, cl client.Client, machin
+// Unlike WaitForCAPIMachinesRunning, this function does not fail the test when machines cannot be provisioned due to insufficient capacity.
+// It returns an error only when machines fail due to insufficient cloud provider capacity, allowing the caller to retry with different configurations.
+func WaitForCAPIMachinesRunningWithRetry(ctx context.Context, cl client.Client, name string, capacityErrorKeys []string) error {
+	machineSet, err := GetCAPIMachineSet(ctx, cl, name)
+	Expect(err).ToNot(HaveOccurred(), "Failed to get CAPI machineset %s", name)
+
+	// Retry until the MachineSet is ready.
+	return wait.PollUntilContextTimeout(ctx, RetryMedium, WaitLong, true, func(ctx context.Context) (bool, error) {
+		machines, err := GetCAPIMachinesFromMachineSet(ctx, cl, machineSet)
+		if err != nil {
+			return false, fmt.Errorf("error getting machines from CAPI machineSet %s: %w", machineSet.Name, err)
+		}
+
+		replicas := ptr.Deref(machineSet.Spec.Replicas, 0)
+		if len(machines) != int(replicas) {
+			klog.Infof("%q: found %d Machines, but MachineSet has %d replicas", name, len(machines), int(replicas))
+			return false, nil
+		}
+
+		// Check for machines with actual failed state (not capacity issues)
+		failed := FilterCAPIMachinesInPhase(machines, string(clusterv1.MachinePhaseFailed))
+		if len(failed) > 0 {
+			return false, handleFailedCAPIMachines(failed)
+		}
+
+		// Check if any machine did not get provisioned because of insufficient capacity.
+		// Check the InfraMachine status for capacity error messages
+		for _, m := range machines {
+			insufficientCapacityResult, insufficientCapacityMessage, err := HasCAPIInsufficientCapacity(ctx, cl, m, capacityErrorKeys)
+			if err != nil {
+				return false, fmt.Errorf("error checking if CAPI machine %s has insufficient capacity: %w", m.Name, err)
+			}
+
+			if insufficientCapacityResult {
+				return false, fmt.Errorf("%w: %s", ErrMachineNotProvisionedInsufficientCloudCapacity, insufficientCapacityMessage)
+			}
+		}
+
+		running := FilterCAPIMachinesInPhase(machines, string(clusterv1.MachinePhaseRunning))
+		// This could probably be smarter, but seems fine for now.
+		if len(running) != len(machines) {
+			klog.Infof("%q: not all CAPI Machines are running: %d of %d", name, len(running), len(machines))
+			return false, nil
+		}
+
+		for _, m := range running {
+			node, err := GetCAPINodeForMachine(ctx, cl, m)
+			if err != nil {
+				klog.Infof("Node for CAPI machine %s not found yet: %v", m.Name, err)
+				return false, nil
+			}
+
+			if !IsNodeReady(node) {
+				klog.Infof("%s: node is not ready", node.Name)
+				return false, nil
+			}
+		}
+
+		return true, nil
+	})
+}
+
+// GetCAPIInfraMachine retrieves the InfraMachine object for the given CAPI machine.
+func GetCAPIInfraMachine(ctx context.Context, cl client.Client, m *clusterv1.Machine) (*unstructured.Unstructured, error) {
```

Comment:
nit: we may want a comment explaining why we return `*unstructured.Unstructured`, and how to use the output of the function :)